packages/template-retail-react-app/app/utils/url.js

```diff
@@ -150,22 +151,77 @@ export const getPathWithLocale = (shortCode, opts = {}) => {
-                    : siteRef || defaultSite.alias || defaultSite.id,
-            locale: isDefaultLocaleOfDefaultSite && isDefaultSite && isHomeRef ? '' : shortCode
-        },
-        opts
+        // By default, as for home page, when the values of site and locale belongs to the default site,
+        // they will be not shown in the url just
+        isDefaultLocaleOfDefaultSite && isDefaultSite && isHomeRef
+            ? ''
+            : siteRef || defaultSite.alias || defaultSite.id,
+        isDefaultLocaleOfDefaultSite && isDefaultSite && isHomeRef ? '' : shortCode
     )
     return newUrl
 }
 
+/**
+ * Generates the URL Template literal (Template string) used to build URLs in the App according
+ * the current selected site/locale and the default App URL configuration.
+ *
+ * @param appConfig Application default configuration.
+ * @param siteUrl Current selected Site object.
+ * @param localeUrl Current selected Locale object.
+ * @returns {function(*, *, *, *=): string} function providing href, site and locale generates a site URL.
+ */
+export const getUrlTemplateLiteral = (appConfig, siteUrl, localeUrl) => {
+    const {site: siteConfig, locale: localeConfig, showDefaults: showDefaultsConfig} = appConfig.url
+    const defaultSite = getDefaultSite()
+    const sites = getSites()
+    const site =
+        sites.find((site) => {
+            return site.alias === appConfig['site'] || site.id === appConfig['site']
+        }) || defaultSite
+    const defaultLocale = getLocaleByReference(site, site.l10n.defaultLocale)
+
+    const isDefaultSite = defaultSite === siteUrl
+    const isDefaultLocale = defaultLocale === localeUrl
+
+    const querySite =
+        (siteConfig === urlPartPositions.QUERY_PARAM && showDefaultsConfig) ||
+        (siteConfig === urlPartPositions.QUERY_PARAM && !showDefaultsConfig && !isDefaultSite)
+    const queryLocale =
+        (localeConfig === urlPartPositions.QUERY_PARAM && showDefaultsConfig) ||
+        (localeConfig === urlPartPositions.QUERY_PARAM && !showDefaultsConfig && !isDefaultLocale)
+
+    const isQuery = querySite || queryLocale
+    const isQuerySiteAndLocale = querySite && queryLocale
+
+    const pathSite =
+        (siteConfig === urlPartPositions.PATH && showDefaultsConfig) ||
+        (siteConfig === urlPartPositions.PATH && !showDefaultsConfig && !isDefaultSite)
+    const pathLocale =
+        (localeConfig === urlPartPositions.PATH && showDefaultsConfig) ||
+        (localeConfig === urlPartPositions.PATH && !showDefaultsConfig && !isDefaultLocale)
+
+    const isPathSiteAndLocale = pathSite && pathLocale
+
+    const templateConfig = {
+        pathSite: pathSite,
+        pathLocale: pathLocale,
+        isPathSiteAndLocale: isPathSiteAndLocale,
+        isQuery: isQuery,
+        isQuerySiteAndLocale: isQuerySiteAndLocale,
+        querySite: querySite,
+        queryLocale: queryLocale
+    }
+
+    return (href, site, locale, c = templateConfig) =>
```

Comment:
Dumb question: How to use the `c` here?

Reply:
`c` is an optional parameter. If not specified, it will default to `templateConfig`, defined above. Would probably be more clear to call it `config` or something.

Reply:
I am trying to think in which scenario I would need to pass a custom `templateConfig` object. 🤔 . Do we to document it?

Reply:
As @wjhsf  mentions, `c` is an optional parameter with the URL template configuration values we use to build the App URL template. 
The idea is to only run the "expensive" methods to obtain the URL template configuration values from the App default configuration only once and share it across the App via React.Context.
The parameter `c` it's not meant to be changed by users, thus the cryptic name `c`. I renamed `c` to `config` to clarify.